README.md
# Carnalyser
An open source search engine to grabble car offers from a wide variety of car trade websites.

## Vision
This project aims to be a new omini channel to deliver insights for users about car purchasing.
Carnalyser goes beyond a simple research result delivering. It gives real insights to support your car purchase, researching across a wide variety of car trade websites, guiding the user to the best option.

## How to Setup Environment
1. Install scrapy (pip install scrapy)
2. Choose a folder and open it using prompt
3. Create a project called Carnalyser by running 'scrapy startproject Carnalyser'
4. Deploy all files inside the folder

## How to run spiders
1. Open prompt and navigate to '.\Carnalyser\spiders' folder
2. Run 'scrapy runspider nameofthespider.py' or
3. Run 'scrapy crawl SpiderName -o SpiderName.json' to get information into a json file


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
[See Issues](../Issues)

## [0.0.1] - 04-17-2018
Initial Release containing only a simple scrapy spider for the WebMotors web site.
### Added
1. CHANGELOG.md file added.
### Changed
none
### Removed
noe
### Deprecated
none


Carnalyser/settings.py
# -*- coding: utf-8 -*-

# Scrapy settings for Carnalyser project
#
# For simplicity, this file contains only settings considered important or
# commonly used. You can find more settings consulting the documentation:
#
#     https://doc.scrapy.org/en/latest/topics/settings.html
#     https://doc.scrapy.org/en/latest/topics/downloader-middleware.html
#     https://doc.scrapy.org/en/latest/topics/spider-middleware.html

BOT_NAME = 'Carnalyser'

SPIDER_MODULES = ['Carnalyser.spiders']
NEWSPIDER_MODULE = 'Carnalyser.spiders'


# Crawl responsibly by identifying yourself (and your website) on the user-agent
#USER_AGENT = 'Carnalyser (+http://www.yourdomain.com)'

# Obey robots.txt rules
ROBOTSTXT_OBEY = False #https://stackoverflow.com/questions/37274835/getting-forbidden-by-robots-txt-scrapy

# Configure maximum concurrent requests performed by Scrapy (default: 16)
#CONCURRENT_REQUESTS = 32

# Configure a delay for requests for the same website (default: 0)
# See https://doc.scrapy.org/en/latest/topics/settings.html#download-delay
# See also autothrottle settings and docs
#DOWNLOAD_DELAY = 3
# The download delay setting will honor only one of:
#CONCURRENT_REQUESTS_PER_DOMAIN = 16
#CONCURRENT_REQUESTS_PER_IP = 16

# Disable cookies (enabled by default)
#COOKIES_ENABLED = False

# Disable Telnet Console (enabled by default)
#TELNETCONSOLE_ENABLED = False

# Override the default request headers:
#DEFAULT_REQUEST_HEADERS = {
#   'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
#   'Accept-Language': 'en',
#}

# Enable or disable spider middlewares
# See https://doc.scrapy.org/en/latest/topics/spider-middleware.html
#SPIDER_MIDDLEWARES = {
#    'Carnalyser.middlewares.CarnalyserSpiderMiddleware': 543,
#}

# Enable or disable downloader middlewares
# See https://doc.scrapy.org/en/latest/topics/downloader-middleware.html
#DOWNLOADER_MIDDLEWARES = {
#    'Carnalyser.middlewares.CarnalyserDownloaderMiddleware': 543,
#}

# Enable or disable extensions
# See https://doc.scrapy.org/en/latest/topics/extensions.html
#EXTENSIONS = {
#    'scrapy.extensions.telnet.TelnetConsole': None,
#}

# Configure item pipelines
# See https://doc.scrapy.org/en/latest/topics/item-pipeline.html
#ITEM_PIPELINES = {
#    'Carnalyser.pipelines.CarnalyserPipeline': 300,
#}

# Enable and configure the AutoThrottle extension (disabled by default)
# See https://doc.scrapy.org/en/latest/topics/autothrottle.html
#AUTOTHROTTLE_ENABLED = True
# The initial download delay
#AUTOTHROTTLE_START_DELAY = 5
# The maximum download delay to be set in case of high latencies
#AUTOTHROTTLE_MAX_DELAY = 60
# The average number of requests Scrapy should be sending in parallel to
# each remote server
#AUTOTHROTTLE_TARGET_CONCURRENCY = 1.0
# Enable showing throttling stats for every response received:
#AUTOTHROTTLE_DEBUG = False

# Enable and configure HTTP caching (disabled by default)
# See https://doc.scrapy.org/en/latest/topics/downloader-middleware.html#httpcache-middleware-settings
#HTTPCACHE_ENABLED = True
#HTTPCACHE_EXPIRATION_SECS = 0
#HTTPCACHE_DIR = 'httpcache'
#HTTPCACHE_IGNORE_HTTP_CODES = []
#HTTPCACHE_STORAGE = 'scrapy.extensions.httpcache.FilesystemCacheStorage'


Carnalyser/middlewares.py
# -*- coding: utf-8 -*-

# Define here the models for your spider middleware
#
# See documentation in:
# https://doc.scrapy.org/en/latest/topics/spider-middleware.html

from scrapy import signals


class CarnalyserSpiderMiddleware(object):
    # Not all methods need to be defined. If a method is not defined,
    # scrapy acts as if the spider middleware does not modify the
    # passed objects.

    @classmethod
    def from_crawler(cls, crawler):
        # This method is used by Scrapy to create your spiders.
        s = cls()
        crawler.signals.connect(s.spider_opened, signal=signals.spider_opened)
        return s

    def process_spider_input(self, response, spider):
        # Called for each response that goes through the spider
        # middleware and into the spider.

        # Should return None or raise an exception.
        return None

    def process_spider_output(self, response, result, spider):
        # Called with the results returned from the Spider, after
        # it has processed the response.

        # Must return an iterable of Request, dict or Item objects.
        for i in result:
            yield i

    def process_spider_exception(self, response, exception, spider):
        # Called when a spider or process_spider_input() method
        # (from other spider middleware) raises an exception.

        # Should return either None or an iterable of Response, dict
        # or Item objects.
        pass

    def process_start_requests(self, start_requests, spider):
        # Called with the start requests of the spider, and works
        # similarly to the process_spider_output() method, except
        # that it doesn’t have a response associated.

        # Must return only requests (not items).
        for r in start_requests:
            yield r

    def spider_opened(self, spider):
        spider.logger.info('Spider opened: %s' % spider.name)


class CarnalyserDownloaderMiddleware(object):
    # Not all methods need to be defined. If a method is not defined,
    # scrapy acts as if the downloader middleware does not modify the
    # passed objects.

    @classmethod
    def from_crawler(cls, crawler):
        # This method is used by Scrapy to create your spiders.
        s = cls()
        crawler.signals.connect(s.spider_opened, signal=signals.spider_opened)
        return s

    def process_request(self, request, spider):
        # Called for each request that goes through the downloader
        # middleware.

        # Must either:
        # - return None: continue processing this request
        # - or return a Response object
        # - or return a Request object
        # - or raise IgnoreRequest: process_exception() methods of
        #   installed downloader middleware will be called
        return None

    def process_response(self, request, response, spider):
        # Called with the response returned from the downloader.

        # Must either;
        # - return a Response object
        # - return a Request object
        # - or raise IgnoreRequest
        return response

    def process_exception(self, request, exception, spider):
        # Called when a download handler or a process_request()
        # (from other downloader middleware) raises an exception.

        # Must either:
        # - return None: continue processing this exception
        # - return a Response object: stops process_exception() chain
        # - return a Request object: stops process_exception() chain
        pass

    def spider_opened(self, spider):
        spider.logger.info('Spider opened: %s' % spider.name)


Carnalyser/spiders/trovitspider.py
# -*- coding: utf-8 -*-
import scrapy
from scrapy.spiders import CrawlSpider, Rule
from scrapy.linkextractors import LinkExtractor
from lxml import html
import numpy as np

class TrovitSpider(CrawlSpider):
    name = "trovitspider"
    allowed_domains = ['carros.trovit.com.br']

    start_urls = [
        'https://carros.trovit.com.br/index.php/cod.search_cars/what_d.honda city/isUserSearch.1'
    ]

    rules = ( Rule(LinkExtractor(allow=('course-finder', ),restrict_xpaths=('//a[@class="paginationResult.next.radius-3.brd.brd2.brd-gray-dark.mrg-gutter-l"]',)), callback='parse_items',follow=True), )


    def parse(self, response):
        i = 0
        for car in response.xpath('//div[contains(@itemtype, "http://schema.org/Car https://schema.org/Offer")]'):

            ## Extracting anf formating the link

            # Extracting and formating the brand and model and removing blank spaces (source format: u' HONDA CITY ')
            temp = car.css('div.details h4 a strong::text').extract()
            brand = temp[0]
            model = temp[1]
            
            # Extracting and formatting the manufactory year and model year (Sorce format: u' yyyy/yyyy')
            #temp = car.css('div.info div.features div.info-veiculo span.content-info-veiculo div.info-veiculo-detalhe::text')[0].extract().strip().split("/")
            #year_manufactory = temp[0]
            #year_model = temp[1]

            # Extracting and formatting the price (Source format [u' ',u'xx.xxx'])
            price = car.css('div.features div.price span.amount::text').extract_first()

            ## Extracting and formatting the version
            # Variable Initialization
            cylinder = 0
            engine = 0
            doors = 0
            #print("price " + price)

            # TODO: This certanly can be better. Search for Numpy to deal with that.
            description = car.css('div.details h4 a::text').extract()[1]
            print(description)
            infos = description.split(" ")
            for desc in infos:
                if desc == '1.0' or desc == '1.4' or desc == '1.5' or desc == '1.6' or desc == '1.8' or desc == '2.0':
                    cylinder = desc
                    pass
                if desc == '8v' or desc == '16v':
                    engine = desc
                    pass
                if desc == '2p' or desc == '4p':
                    doors = desc
                    pass
                if desc == 'flex' or desc == 'gasolina':
                    fuel_type = desc
                    pass
                pass 

            ## Extracting and formatting kilomiters
            #km = car.css('div.info div.features div.info-veiculo span.content-info-veiculo div.info-veiculo-detalhe::text')[1].extract().replace(' ','').replace('km','')
            
            #gear_shift = car.css('div.info div.features div.info-veiculo span.content-info-veiculo div.info-veiculo-detalhe::text')[2].extract().strip()
            
            desc = ''.join(car.css('div.details div.description p::text').extract())
            ## Among the values
            yield {
                'brand': brand,
                'model': model,
                'price': price,
                #'year_manufactory': year_manufactory,
                #'year_model': year_model,
                'cylinder': cylinder,
                'engine': engine,
                'doors': doors,
                'fuel': fuel_type,
                #'km': km,
                #'cambio': gear_shift,
                'desc': desc
           }


Carnalyser/spiders/wmspider.py
# -*- coding: utf-8 -*-
import scrapy
from scrapy.spiders import CrawlSpider, Rule
from scrapy.loader import ItemLoader
from scrapy.linkextractors import LinkExtractor
from lxml import html
from .items import Car

class WebMotorsSpider(CrawlSpider):
    name = "wmspider"
    allowed_domains = ['www.webmotors.com.br']

    start_urls = [
        'https://www.webmotors.com.br/carros/sp-osasco/honda/city',
    ]

    rules = ( Rule(LinkExtractor(allow=('course-finder', ),restrict_xpaths=('//a[@class="paginationResult.next.radius-3.brd.brd2.brd-gray-dark.mrg-gutter-l"]',)), callback='parse_items',follow=True), )

    def parse(self, response):
        for car in response.xpath('//a[contains(@itemtype, "http://schema.org/Offer")]'):
            ## Extracting and formating the link

            #using ItemLoader to put info inside it
            i = ItemLoader(Car(),car)
            # Extracting and formating the brand and model and removing blank spaces (source format: u' HONDA CITY ')
            temp = car.css('div.info span.make-model-financiamento::text').extract_first().strip().split(" ")
            brand = temp[0]
            model = temp[1]
            i.add_value('brand',brand)
            i.add_value('model',model)

            # Extracting and formatting the manufactory year and model year (Sorce format: u' yyyy/yyyy')
            temp = car.css('div.info div.features div.info-veiculo span.content-info-veiculo div.info-veiculo-detalhe::text')[0].extract().strip().split("/")
            year_manufactory = temp[0]
            year_model = temp[1]
            i.add_value('year_manufactory',year_manufactory)
            i.add_value('year_model',year_model)

            # Extracting and formatting the price (Source format [u' ',u'xx.xxx'])
            price = (''.join(car.css('div.info div.price-novo.space-preco::text').extract())).strip()
            i.add_value('price',price)

            ## Extracting and formatting the version
            # Variable Initialization
            cylinder = 0
            engine = 0
            doors = 0

            # TODO: This certanly can be better. Search for Numpy to deal with that.
            description = car.css('div.info span.version::text').extract_first().strip()
            infos = description.split(" ")
            for desc in infos:
                if desc == '1.0' or desc == '1.4' or desc == '1.5' or desc == '1.6' or desc == '1.8' or desc == '2.0':
                    cylinder = desc
                    i.add_value('cylinder',cylinder)
                    pass
                if desc == '8V' or desc == '16V':
                    engine = desc
                    i.add_value('engine',engine)
                    pass
                if desc == '2P' or desc == '4P':
                    doors = desc
                    i.add_value('doors',doors)
                    pass
                if desc == 'FLEX' or desc == 'GASOLINA':
                    fuel_type = desc
                    i.add_value('fuel',fuel_type)
                    pass
                pass 

            ## Extracting and formatting kilomiters
            km = car.css('div.info div.features div.info-veiculo span.content-info-veiculo div.info-veiculo-detalhe::text')[1].extract().replace(' ','').replace('km','')
            i.add_value('km',km)
            
            gear_shift = car.css('div.info div.features div.info-veiculo span.content-info-veiculo div.info-veiculo-detalhe::text')[2].extract().strip()
            i.add_value('cambio',gear_shift)
            
            ## Among the values
            yield i.load_item()
            """yield {
                'brand': brand,
                'model': model,
                'price': price,
                'year_manufactory': year_manufactory,
                'year_model': year_model,
                'cylinder': cylinder,
                'engine': engine,
                'doors': doors,
                'fuel': fuel_type,
                'km': km,
                'cambio': gear_shift,
                'desc': description
                # See https://stackoverflow.com/questions/43922562/scrapy-how-to-use-items-in-spider-and-how-to-send-items-to-pipelines
           } """
